src/PowerShellEditorServices/Workspace/Workspace.cs

```diff
@@ -292,67 +299,106 @@ public IEnumerable<string> EnumeratePSFiles()
 
         #region Private Methods
 
+        /// <summary>
+        /// Find PowerShell files recursively down from a given directory path.
+        /// Currently returns files in depth-first order.
+        /// Directory.GetFiles(folderPath, pattern, SearchOption.AllDirectories) would provide this,
+        /// but a cycle in the filesystem will cause that to enter an infinite loop.
+        /// </summary>
+        /// <param name="folderPath">The absolute path of the base folder to search.</param>
+        /// <returns>
+        /// All PowerShell files in the recursive directory hierarchy under the given base directory, up to 64 directories deep.
+        /// </returns>
         private IEnumerable<string> RecursivelyEnumerateFiles(string folderPath)
         {
-            var foundFiles = Enumerable.Empty<string>();
-            var patterns = new string[] { @"*.ps1", @"*.psm1", @"*.psd1" };
+            var foundFiles = new List<string>();
+            var dirStack = new Stack<string>();
```

Comment:
Not sure which is faster, but have you considered a pattern like this:

```csharp
public static IEnumerable<string> RecursiveEnumerateFiles(string filePath)
{
    var results = new List<string>();
    AddEnumeratedFiles(results, filePath, 0, 50);
    return results;
}

private static void AddEnumeratedFiles(List<string> files, string filePath, int currentDepth, int maxDepth)
{
    files.AddRange(Directory.GetFiles(filePath, "*.ps1", SearchOption.TopDirectoryOnly));
    if (currentDepth >= maxDepth)
    {
        return;
    }

    foreach (string directory in Directory.GetDirectories(filePath))
    {
        AddEnumeratedFiles(files, directory, currentDepth + 1, maxDepth);
    }
}
```

That would remove the need for the stack. That said, I don't know enough about the performance implications of using that much recursion (assuming the second method isn't inlined into a loop in the first method?).

Either way, your solution is very clever :) I wouldn't have thought of doing it that way.

Reply:
the implementation @rjmholt implemented is a classic Computer Science problem (which is probably one of the first times my degree has helped!) 

https://www.geeksforgeeks.org/iterative-depth-first-traversal/

you can do it recursively... but a lot of people find iterative to be more readable.

Reply:
Well the big difference is that a recursive call implicitly uses .NET's call stack. It should be more performant than my manual stack usage, but a `StackOverflowException` is uncatchable.

My original re-implementation was almost identical to yours @SeeminglyScience, but the fact that we got a `StackOverflowException` here and being able to turn it into a `while` loop sort of convinced me to just switch it over to the "manual recursive" way.

I actually also looked up whether .NET will do tail call optimisation, and it can, but the C# compiler doesn't emit the right bytecode (F#'s does). So as far as I can tell, `AddEnumeratedFiles` wouldn't get inlined.